src/adsp_config.c

```diff
@@ -0,0 +1,1158 @@
+	out->write_firmware_meu = man_write_fw_meu_v1_8;
+
+	ret = parse_adsp(toml, &ctx, out);
+	if (ret < 0)
+		return err_key_parse("adsp");
+
+	ret = parse_mem_cse(toml, &ctx, &out->man_v1_8->cse_partition_dir_header,
+			    out->man_v1_8->cse_partition_dir_entry);
+	if (ret < 0)
+		return err_key_parse("cse");
+
+	ret = parse_mem_css_v1_8(toml, &ctx, &out->man_v1_8->css);
+	if (ret < 0)
+		return err_key_parse("css");
+
+	ret = parse_mem_signed_pkg(toml, &ctx, &out->man_v1_8->signed_pkg);
+	if (ret < 0)
+		return err_key_parse("signed_pkg");
+
+	ret = parse_mem_partition_info_ext(toml, &ctx, &out->man_v1_8->partition_info);
+	if (ret < 0)
+		return err_key_parse("partition_info");
+
+	ret = parse_mem_adsp_file_ext_v1_8(toml, &ctx, &out->man_v1_8->adsp_file_ext);
+	if (ret < 0)
+		return err_key_parse("adsp_file");
+
+	ret = parse_mem_fw_desc(toml, &ctx, &out->man_v1_8->desc);
+	if (ret < 0)
+		return err_key_parse("fw_desc");
+
+	return 0;
+}
+
+static int parse_adsp_config_v2_5(const toml_table_t *toml, struct adsp *out)
+{
+	struct parse_ctx ctx;
+	int ret;
+
+	out->man_v2_5 = malloc(sizeof(struct fw_image_manifest_v2_5));
+	if (!out->man_v2_5)
+		return err_malloc("man_v2_5");
+
+	out->write_firmware = NULL;
+	out->write_firmware_meu = man_write_fw_meu_v2_5;
+
+	ret = parse_adsp(toml, &ctx, out);
+	if (ret < 0)
+		return err_key_parse("adsp");
+
+	ret = parse_mem_cse(toml, &ctx, &out->man_v2_5->cse_partition_dir_header,
+			    out->man_v2_5->cse_partition_dir_entry);
+	if (ret < 0)
+		return err_key_parse("cse");
+
+	ret = parse_mem_css_v1_8(toml, &ctx, &out->man_v2_5->css);
+	if (ret < 0)
+		return err_key_parse("css");
+
+	ret = parse_mem_signed_pkg(toml, &ctx, &out->man_v2_5->signed_pkg);
+	if (ret < 0)
+		return err_key_parse("signed_pkg");
+
+	ret = parse_mem_partition_info_ext(toml, &ctx, &out->man_v2_5->partition_info);
+	if (ret < 0)
+		return err_key_parse("partition_info");
+
+	ret = parse_mem_adsp_file_ext_v2_5(toml, &ctx, &out->man_v2_5->adsp_file_ext);
+	if (ret < 0)
+		return err_key_parse("adsp_file");
+
+	ret = parse_mem_fw_desc(toml, &ctx, &out->man_v2_5->desc);
+	if (ret < 0)
+		return err_key_parse("fw_desc");
+
+	return 0;
+}
+
+static inline bool check_config_version(int major, int minor, const int64_t *version)
+{
+	return version[0] == major && version[1] == minor;
+}
+
+static int parse_adsp_config_fd(FILE *fd, struct adsp *out)
+{
+	int64_t css_version[2];
+	toml_table_t* toml;
+	toml_array_t* arr;
+	toml_raw_t raw;
+	char errbuf[256];
+	int ret;
+	int i;
+
+	toml = toml_parse_file(fd, errbuf, ARRAY_SIZE(errbuf));
+	if (!toml)
+		return log_err(-EINVAL, "error: toml file parsing\n");
+
+	/* check "version" key */
+	arr = toml_array_in(toml, "version");
+	if (!arr || toml_array_type(arr) != 'i'  || toml_array_nelem(arr) != 2 ||
+	    toml_array_kind(arr) != 'v') {
+		ret = log_err(-EINVAL, "error: toml parse key 'version' failed\n");
+		goto error;
+	}
+
+	for(i = 0; i < ARRAY_SIZE(css_version); ++i){
```

Comment:
Missing space after for(

Reply:
Added, btw shouldn't we add checkpach to PR in this repo ?

Reply:
Sounds like a good idea